File: sdk/python/aragora_sdk/namespaces/workflow_templates.py
"""
Workflow Templates Namespace API

Provides access to pre-built workflow templates for common automation patterns:
- List available workflow templates
- Get template details
- Get full template packages with definitions and examples
- Run templates with inputs
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient

class WorkflowTemplatesAPI:
    """
    Synchronous Workflow Templates API.

    Provides access to pre-built workflow templates for common automation patterns.

    Example:
        >>> client = AragoraClient(base_url="https://api.aragora.ai")
        >>> templates = client.workflow_templates.list(category="analysis")
        >>> for template in templates["templates"]:
        ...     print(template["name"], template["description"])
        >>> # Run a template
        >>> result = client.workflow_templates.run(
        ...     template_id="tpl_123",
        ...     inputs={"document": "contract.pdf"}
        ... )
        >>> print(result["execution_id"])
    """

    def __init__(self, client: AragoraClient):
        self._client = client

    def list(
        self,
        category: str | None = None,
        pattern: str | None = None,
        search: str | None = None,
        tags: list[str] | None = None,
        limit: int = 50,
        offset: int = 0,
    ) -> dict[str, Any]:
        """
        List available workflow templates.

        Args:
            category: Filter by category (e.g., "analysis", "review", "automation")
            pattern: Filter by pattern type
            search: Search query for template name/description
            tags: Filter by tags
            limit: Maximum number of templates to return (default: 50)
            offset: Pagination offset (default: 0)

        Returns:
            Dictionary with 'templates' list and 'total' count
        """
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        if category:
            params["category"] = category
        if pattern:
            params["pattern"] = pattern
        if search:
            params["search"] = search
        if tags:
            params["tags"] = tags

        return self._client.request("GET", "/api/v1/workflow/templates", params=params)

    def get(self, template_id: str) -> dict[str, Any]:
        """
        Get a specific workflow template by ID.

        Args:
            template_id: The template ID

        Returns:
            Template details including id, name, description, category,
            pattern, tags, version, and timestamps
        """
        return self._client.request("GET", f"/api/v1/workflow/templates/{template_id}")

    def get_package(self, template_id: str) -> dict[str, Any]:
        """
        Get the full template package including definition and examples.

        Args:
            template_id: The template ID

        Returns:
            Template package including:
            - template: Basic template metadata
            - definition: Full workflow definition
            - dependencies: Required dependencies
            - examples: Usage examples with sample inputs
        """
        return self._client.request("GET", f"/api/v1/workflow/templates/{template_id}/package")


class AsyncWorkflowTemplatesAPI:
    """
    Asynchronous Workflow Templates API.

    Provides access to pre-built workflow templates for common automation patterns.

    Example:
        >>> async with AragoraAsyncClient(base_url="https://api.aragora.ai") as client:
        ...     templates = await client.workflow_templates.list(category="analysis")
        ...     for template in templates["templates"]:
        ...         print(template["name"], template["description"])
        ...     # Run a template
        ...     result = await client.workflow_templates.run(
        ...         template_id="tpl_123",
        ...         inputs={"document": "contract.pdf"}
        ...     )
        ...     print(result["execution_id"])
    """

    def __init__(self, client: AragoraAsyncClient):
        self._client = client

    async def list(
        self,
        category: str | None = None,
        pattern: str | None = None,
        search: str | None = None,
        tags: list[str] | None = None,
        limit: int = 50,
        offset: int = 0,
    ) -> dict[str, Any]:
        """
        List available workflow templates.

        Args:
            category: Filter by category (e.g., "analysis", "review", "automation")
            pattern: Filter by pattern type
            search: Search query for template name/description
            tags: Filter by tags
            limit: Maximum number of templates to return (default: 50)
            offset: Pagination offset (default: 0)

        Returns:
            Dictionary with 'templates' list and 'total' count
        """
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        if category:
            params["category"] = category
        if pattern:
            params["pattern"] = pattern
        if search:
            params["search"] = search
        if tags:
            params["tags"] = tags

        return await self._client.request("GET", "/api/v1/workflow/templates", params=params)

    async def get(self, template_id: str) -> dict[str, Any]:
        """
        Get a specific workflow template by ID.

        Args:
            template_id: The template ID

        Returns:
            Template details including id, name, description, category,
            pattern, tags, version, and timestamps
        """
        return await self._client.request("GET", f"/api/v1/workflow/templates/{template_id}")

    async def get_package(self, template_id: str) -> dict[str, Any]:
        """
        Get the full template package including definition and examples.

        Args:
            template_id: The template ID

        Returns:
            Template package including:
            - template: Basic template metadata
            - definition: Full workflow definition
            - dependencies: Required dependencies
            - examples: Usage examples with sample inputs
        """
        return await self._client.request(
            "GET", f"/api/v1/workflow/templates/{template_id}/package"
        )

File: sdk/typescript/src/namespaces/workflow-templates.ts
/**
 * Workflow Templates Namespace API
 *
 * Provides access to pre-built workflow templates for common automation patterns.
 */

export interface WorkflowTemplate {
  id: string;
  name: string;
  description?: string;
  category: string;
  pattern?: string;
  tags?: string[];
  version?: string;
  created_at?: string;
  updated_at?: string;
  metadata?: Record<string, unknown>;
}

export interface WorkflowTemplatePackage {
  template: WorkflowTemplate;
  definition: Record<string, unknown>;
  dependencies?: string[];
  examples?: Array<{ name: string; inputs: Record<string, unknown> }>;
}

export interface WorkflowTemplateRunResult {
  execution_id: string;
  status: 'started' | 'running' | 'completed' | 'failed';
  template_id: string;
  started_at: string;
  result?: Record<string, unknown>;
}

export interface ListWorkflowTemplatesParams {
  category?: string;
  pattern?: string;
  search?: string;
  tags?: string[];
  limit?: number;
  offset?: number;
}

export interface RunWorkflowTemplateParams {
  inputs?: Record<string, unknown>;
  config?: Record<string, unknown>;
  workspace_id?: string;
}

interface WorkflowTemplatesClientInterface {
  request<T = unknown>(
    method: string,
    path: string,
    options?: { params?: Record<string, unknown>; body?: unknown }
  ): Promise<T>;
}

export class WorkflowTemplatesAPI {
  constructor(private client: WorkflowTemplatesClientInterface) {}

  /**
   * List available workflow templates.
   */
  async list(
    params?: ListWorkflowTemplatesParams
  ): Promise<{ templates: WorkflowTemplate[]; total: number }> {
    return this.client.request('GET', '/api/v1/workflow/templates', {
      params: params as Record<string, unknown>,
    });
  }

  /**
   * Get a specific workflow template by ID.
   */
  async get(templateId: string): Promise<WorkflowTemplate> {
    return this.client.request('GET', `/api/v1/workflow/templates/${templateId}`);
  }

  /**
   * Get the full template package including definition and examples.
   */
  async getPackage(templateId: string): Promise<WorkflowTemplatePackage> {
    return this.client.request('GET', `/api/v1/workflow/templates/${templateId}/package`);
  }

  /**
   * Run a workflow template with the given inputs.
   */
  async run(
    templateId: string,
    params?: RunWorkflowTemplateParams
  ): Promise<WorkflowTemplateRunResult> {
    return this.client.request('POST', `/api/v1/workflow/templates/${templateId}/run`, {
      body: params,
    });
  }

  /**
   * List workflow executions.
   */
  async listExecutions(params?: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.client.request('GET', '/api/v1/workflows/executions', { params }) as Promise<Record<string, unknown>>;
  }

  /**
   * Get workflow execution.
   */
  async getExecution(executionId: string): Promise<Record<string, unknown>> {
    return this.client.request('GET', `/api/v1/workflows/executions/${executionId}`) as Promise<Record<string, unknown>>;
  }

  /**
   * List workflow templates.
   */
  async listTemplates(params?: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.client.request('GET', '/api/v1/workflows/templates', { params }) as Promise<Record<string, unknown>>;
  }

  /**
   * Get workflow template.
   */
  async getTemplate(templateId: string): Promise<Record<string, unknown>> {
    return this.client.request('GET', `/api/v1/workflows/templates/${templateId}`) as Promise<Record<string, unknown>>;
  }
}
